### PR TITLE
Remove some needless allocations

### PR DESCRIPTION
**Pull request by elichai**
Hi,
I removed allocations in places that we serialized into a Vec just to then hash it. instead we can feed the hasher itself instead of a vector and serialize into that.

### CODE REVIEW

**Review by dpc — DISMISSED** [review]

File: src/blockdata/transaction.rs [hunk 1]
```diff
@@ -378,9 +378,12 @@ impl Transaction {
             _ => unreachable!()
         };
         // hash the result
-        let mut raw_vec = serialize(&tx);
-        raw_vec.extend_from_slice(&endian::u32_to_array_le(sighash_u32));
-        SigHash::hash(&raw_vec)
+        // TODO: Sanity assert that consensus_encode returned length matches the hashed length in the hasher.
```
Comment (stevenroose) [inline src/blockdata/transaction.rs:381]: I don't think that's necessary.

File: src/blockdata/transaction.rs [hunk 2]
```diff
@@ -378,9 +378,12 @@ impl Transaction {
...
+        let sighash_arr = endian::u32_to_array_le(sighash_u32);
```
Comment (stevenroose) [inline src/blockdata/transaction.rs:382]: Could you move this line down right above to where the variable is used?

File: src/util/misc.rs [hunk 3]
```diff
@@ -59,14 +59,15 @@ pub fn script_find_and_remove(haystack: &mut Vec<u8>, needle: &[u8]) -> usize {
...
 pub fn signed_msg_hash(msg: &str) -> sha256d::Hash {
-    sha256d::Hash::hash(
-        &[
-            MSG_SIGN_PREFIX,
-            &encode::serialize(&encode::VarInt(msg.len() as u64)),
-            msg.as_bytes(),
-        ]
-        .concat(),
-    )
+    let msg_len = encode::VarInt(msg.len() as u64);
```
Comment (stevenroose) [inline src/util/misc.rs:62]: Again, is there a reason for this to be here instead of right before where it's inputted?

File: src/util/misc.rs [hunk 4]
```diff
@@ -59,14 +59,15 @@ pub fn script_find_and_remove(haystack: &mut Vec<u8>, needle: &[u8]) -> usize {
...
+
+    // TODO: Sanity assert that consensus_encode returned length matches the hashed length in the hasher.
```
Comment (stevenroose) [inline src/util/misc.rs:64]: Again IMO not needed.

**Review by stevenroose — COMMENTED** [review]
Very much concept ACK. I remember changing many of the `Hash::hash` usages into consensus encoding into the engine as well. All these are such obvious wins :)

File: src/blockdata/transaction.rs (hunk 1 above)
Comment (elichai) [reply src/blockdata/transaction.rs]: You're right. I might be over-paranoid

File: src/blockdata/transaction.rs (hunk 2 above)
Comment (elichai) [reply src/blockdata/transaction.rs]: Done

File: src/util/misc.rs (hunk 3 above)
Comment (elichai) [reply src/util/misc.rs]: Done

**Review by stevenroose — DISMISSED** [review]
utACK 83e90777ac644e54d308f2b0d4c194bf6a0e2c38

**Comment by stevenroose** [comment]
I restarted the nightly test that timed out. I'm pretty confident that it will be ok.

**Comment by elichai** [comment]
Travis passed but github doesn't recognize it for some reason, trying to close and re-open to trigger both github and travis

**Review by sgeisler — DISMISSED** [review]

**Comment by sgeisler** [comment]
Maybe you can just git amend with an empty change set and force push to trigger travis? It would invalidate the reviews but that's not a big deal since we could just look at the diff and re-ACK if it's none.

**Comment by elichai** [comment]
I force pushed an empty amend because github's status checks are really stubborn, closing-reopening didn't work.

@sgeisler @stevenroose Could you please re-approve?

**Review by stevenroose — APPROVED** [review]
LGTM

**Review by sgeisler — APPROVED** [review]

**Pull request merged**